File: 0387-first-unique-character-in-a-string/0387-first-unique-character-in-a-string.cpp
class Solution {
public:
    int firstUniqChar(string s) {
        std::unordered_map<char, int> charFrequency;

    // Populate the frequency map
    for (char c : s) {
        charFrequency[c]++;
    }

    // Find the first character with a frequency of 1
    for (int i = 0; i < s.length(); ++i) {
        if (charFrequency[s[i]] == 1) {
            return i;
        }
    }

    // If no unique character is found, return -1
    return -1;
    }
};

File: 0387-first-unique-character-in-a-string/0387-first-unique-character-in-a-string.rs
use std::collections::HashMap;

impl Solution {
    pub fn first_uniq_char(s: String) -> i32 {
        let mut char_frequency = HashMap::new();

    // Populate the frequency map
    for char in s.chars() {
        *char_frequency.entry(char).or_insert(0) += 1;
    }

    // Find the first character with a frequency of 1
    for (index, char) in s.chars().enumerate() {
        if char_frequency[&char] == 1 {
            return index as i32;
        }
    }

    // If no unique character is found, return -1
    -1
    }
}